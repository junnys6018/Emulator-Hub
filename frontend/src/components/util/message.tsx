import React, { useCallback, useContext, useEffect, useRef, useState } from 'react';
import { FaTimes } from 'react-icons/fa';

const borderColorMap = {
    SUCCESS: 'border-l-4 border-green-500',
    WARN: 'border-l-4 border-yellow-500',
    ERROR: 'border-l-4 border-red-500',
    INFO: 'border-l-4 border-gray-700',
};

type Severity = 'SUCCESS' | 'INFO' | 'WARN' | 'ERROR';

interface MessageProps {
    title: string;
    message: string;
    severity: Severity;
    close: () => void;
}

export function Message(props: MessageProps) {
    const container = useRef<HTMLDivElement>(null);

    useEffect(() => {
        (container.current as HTMLDivElement).classList.remove('opacity-0');
        (container.current as HTMLDivElement).classList.add('opacity-100');
    }, []);

    return (
        <div
            className={`opacity-0 transition duration-300 flex mb-4 bg-gray-700 ${borderColorMap[props.severity]}`}
            style={{ minHeight: '64px' }}
            ref={container}
        >
            <div className={`w-80 flex flex-col justify-center`}>
                <h5 className="font-semibold ml-3 pt-3">{props.title}</h5>
                <span className="text-sm ml-3 pb-3">{props.message}</span>
            </div>
            <button className="hover:text-red-500 ml-4 self-start mt-3 mr-3" onClick={props.close}>
                <FaTimes size="1.25rem" />
            </button>
        </div>
    );
}

interface MessageProviderProps {
    children: React.ReactNode;
}

interface MessageOptions {
    title: string;
    severity: Severity;
}

export const MessageContext = React.createContext<
    ((message: string, options?: Partial<MessageOptions>) => void) | null
>(null);

export function MessageProvider(props: MessageProviderProps) {
    const [messages, setMessages] = useState<{ title: string; message: string; severity: Severity; key: number }[]>([]);

    const key = useRef(0);
    const timeouts = useRef<{
        [key: string]: number;
    }>({});

    const close = useCallback((key: number) => {
        setMessages(messages => messages.filter(value => value.key !== key));
    }, []);

    const message = useCallback(
        (message: string, options?: Partial<MessageOptions>) => {
            const defaults: MessageOptions = {
                title: 'Message',
                severity: 'INFO',
            };

            options = { ...defaults, ...options };
            const currentKey = key.current++;
            setMessages(messages => messages.concat({ key: currentKey, message, ...(options as MessageOptions) }));
            const timerId = setTimeout(() => {
                close(currentKey);
                delete timeouts.current[currentKey];

                // We need to do this annoying cast becuse typescript is using the function signature of the nodejs version of setTimeout
                // instead of the browsers version
            }, 5000) as unknown as number;
            timeouts.current[currentKey] = timerId;
        },
        [close],
    );

    return (
        <MessageContext.Provider value={message}>
            {props.children}
            <div className="fixed z-40 bottom-4 right-1/2 transform translate-x-1/2 md:right-4 md:translate-x-0 flex flex-col">
                {messages.map(({ title, message, severity, key }) => (
                    <Message
                        title={title}
                        message={message}
                        severity={severity}
                        key={key}
                        close={() => {
                            close(key);
                            clearTimeout(timeouts.current[key]);
                        }}
                    />
                ))}
            </div>
        </MessageContext.Provider>
    );
}

export function useMessage() {
    const message = useContext(MessageContext);
    if (message === null) {
        throw new Error('useMessage must be used within MessageProvider');
    }
    return message;
}
