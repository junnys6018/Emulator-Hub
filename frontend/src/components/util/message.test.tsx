/* eslint-disable @typescript-eslint/no-empty-function */
import React from 'react';
import renderer from 'react-test-renderer';
import { Message } from './message';

test('<Message /> snapshot', () => {
    const tree = renderer
        .create(<Message title="Message" message="This is a message" severity="SUCCESS" close={() => {}} />)
        .toJSON();
    expect(tree).toMatchSnapshot();
});
