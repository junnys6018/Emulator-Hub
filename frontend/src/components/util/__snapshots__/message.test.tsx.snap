// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Message /> snapshot 1`] = `
<div
  className="opacity-0 transition flex mb-4 bg-gray-700 border-l-4 border-green-500"
  style={
    Object {
      "minHeight": "64px",
    }
  }
>
  <div
    className="w-80 flex flex-col justify-center"
  >
    <h1
      className="font-semibold ml-3 pt-3"
    >
      Message
    </h1>
    <span
      className="text-sm ml-3 pb-3"
    >
      This is a message
    </span>
  </div>
  <button
    className="hover:text-red-500 ml-4 self-start mt-3 mr-3"
    onClick={[Function]}
  >
    <svg
      fill="currentColor"
      height="20px"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 352 512"
      width="20px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
      />
    </svg>
  </button>
</div>
`;
